src/MatrixTwitter.js

```diff
@@ -0,0 +1,1040 @@
+              body,
+              response.statusCode
+            );
+      } else {
+        try {
+          var jsonresponse = JSON.parse(body);
+          if (jsonresponse.token_type == "bearer") {
+            fs.writeFile("bearer.tok", jsonresponse.access_token, (err) => {
+              if (err) {
+                            //This error is unfortunate, but not a failure to retrieve a token so the bridge can run fine.
+                log.error("Twitter", "Couldn't write bearer token to file. Reason:", err);
+              }
+            });
+                    //Not waiting for callback since it is trivial to get a new token, and can be done async
+            resolve(jsonresponse.bearer_token);
+          } else {
+            reject("Request to oauth2/post did not return the correct" +
+                    "token type ('bearer'). This is weeeird.");
+            log.error("Twitter", "Body of response:%s", body);
+          }
+        } catch (e) {
+          reject(e);
+        }
+      }
+    });
+  });
+}
+
+MatrixTwitter.prototype._get_bearer_token = function () {
+  return new Promise((resolve) => {
+    fs.readFile('bearer.tok', {encoding: 'utf-8'}, (err, content) => {
+      if(err) {
+        log.warn('Twitter', "Token file not found or unreadable. Requesting new token.");
+        log.error("Twitter", err);
+        resolve(Promise.resolve(this._get_bearer_http()));
+      }
+      resolve(content);
+    });
+  }).then(token => {
+    //Test the token
+    return new Promise((resolve, reject) =>{
+      var auth = {
+        consumer_key: this.app_auth.consumer_key,
+        consumer_secret: this.app_auth.consumer_secret,
+        bearer_token: token
+      };
+      this.app_twitter = new Twitter(auth).get(
+        'application/rate_limit_status',
+        {},
+        (error, status, response) => {
+          if(response.statusCode == 401) {
+            log.warn('Twitter', "Authentication with existing token failed. ");
+            fs.unlink('bearer.tok', (err) => {
+              if(err) {
+                log.warn('Twitter', "Couldn't delete bearer.tok");
+              }
+              resolve(this._get_bearer_http());
+            });
+          }
+          else if (response.statusCode == 200) {
+            log.info('Twitter', "Existing token OK.");
+            resolve(token);
+          }
+          else {
+            log.error("Twitter", error);
+            reject("Unexpected response to application/rate_limit_status " +
+              "during bearer token validation. Bailing.");
+          }
+        });
+    });
+  });
+}
+
+MatrixTwitter.prototype._get_intent = function (id) {
+  return this._bridge.getIntentFromLocalpart("twitter_" + id);
+}
+
+MatrixTwitter.prototype._update_user_timeline_profile = function (profile) {
+  var ts = new Date().getTime();
+  if(profile == null) {
+    log.warn("Twitter", "Tried to preform a profile update with a null profile.");
+    return;
```

Comment:
The function signature to `_update_user_timeline_profile` is horrible. You conditionally return a `Promise` :( either return a Promise always or never. Looking at your usages of this function, I would just not return a `Promise` (you never seem to use it).